### ISSUE TITLE
Separate GREGoR processing

### ISSUE THREAD

**Issue by bwalsh** [gks-anvil/vrs_anvil_toolkit#98]
For discussion:

# Specification: VCF to VRS Processing Package with Cohort Allele Frequency Generation

## Overview
This package processes Variant Call Format (VCF) files to ensure compatibility with VRS (Variation Representation Specification). It checks each line for a VRS identifier, creates one if missing, and processes subject and phenotype data via a plugin architecture. Finally, it generates a Cohort Allele Frequency (CAF) object.

---

### 1. **Read and Parse VCF Files**
- **Input**: Standard VCF file.
- **Functionality**:
  - Validate VCF format and schema compliance.
  - Extract metadata (header) and variants (data lines).
- **Output**: Parsed representation of the VCF file in a data structure for downstream operations.
- [x] Accomplished by vrs_toolkit
---

### 2. **VRS Identifier Validation and Generation**
- **Input**: Parsed VCF line.
- **Functionality**:
  - Check for existing VRS identifiers in each VCF line.
  - Generate a VRS identifier if missing, using:
    - **Input data**: Chromosome, position, reference allele, alternate allele.
    - **Output**: Valid VRS identifier according to VRS schema.
- **Output**: Updated VCF line with VRS identifier.
- [x] Accomplished by vrs_toolkit
---

### 3. **Plugin Architecture for Subject and Phenotype Association**
- **Goal**: Enable flexible integration of subject/phenotype determination methods.
- **Architecture**:
  - Define a plugin interface with:
    - `get_subject(line: dict) -> str`: Identifies the subject for a VCF line.
    - `get_phenotype(line: dict) -> list`: Identifies phenotypes linked to the variant.
  - Allow multiple plugins to run in parallel or series based on configuration. See [python packaging](https://packaging.python.org/en/latest/guides/creating-and-discovering-plugins/#using-naming-convention)
- **Input**: VCF line.
- **Functionality**:
  - Determine the associated subject (e.g., patient, sample).
  - Identify any phenotypes linked to the variant.
- **Output**: Annotated VCF line with subject and phenotype data.
- [ ] Partially implemented by notebook
---

### 4. **Cohort Allele Frequency (CAF) Object Generation**
- **Input**: Processed VCF data.
- **Functionality**:
  - Aggregate allele frequency data across subjects.
  - Generate a CAF object containing:
    - Variant identifier (VRS).
    - Allele frequencies across the cohort.
    - Population metadata (e.g., demographic stratification if available).
- **Output**: CAF object in JSON or another standard format for downstream analysis.
- [ ] Partially implemented by notebook

---

### 1. **Command Line Interface (CLI)**
- **Features**:
  - `read_vcf`: Specify the VCF file to process.
  - `--output`: Path to save the processed VCF with annotations.
  - `--plugins`: List of plugins to use for subject/phenotype determination.
- **Example**:
  ```bash
  vcf_to_vrs --input sample.vcf --output annotated.vcf --plugins plugin1,plugin2

**Comment by bwalsh** [comment]
### 2. **Programmatic Interface (Python API)**

#### Key Functions:
- **`process_vcf(input_path: str, output_path: str, plugins: list) -> None`**  
  Processes a VCF file, ensuring VRS identifiers are added, applying plugins, and saving the updated VCF.  

  **Parameters**:
  - `input_path`: Path to the input VCF file.
  - `output_path`: Path to save the processed VCF.
  - `plugins`: A list of plugin objects to determine subject/phenotype data.

---

- **`generate_caf(vcf_data: list) -> dict`**  
  Generates a Cohort Allele Frequency (CAF) object based on processed VCF data.

  **Parameters**:
  - `vcf_data`: Parsed VCF data in list format.

  **Returns**:
  - A dictionary representing the CAF object, including allele frequencies and population metadata.

---

- **`register_plugin(plugin: Plugin) -> None`**  
  Registers a plugin to extend the system’s capabilities.

  **Parameters**:
  - `plugin`: A plugin object implementing the required interface for subject and phenotype determination.

  **Usage Example**:
  ```python
  from vcf_to_vrs import process_vcf, generate_caf, register_plugin

  # Register a plugin
  register_plugin(my_custom_plugin)

  # Process a VCF file
  process_vcf("input.vcf", "output.vcf", plugins=[my_custom_plugin])

  # Generate a CAF object
  caf_object = generate_caf(parsed_vcf_data)
  print(caf_object)
  ```

**Comment by bwalsh** [comment]
## Example Workflow

1. **Input**: Provide a VCF file with potentially missing VRS identifiers.
2. **Process**:
   - Parse the VCF file to extract metadata and variants.
   - Check each line for existing VRS identifiers and generate any missing ones.
   - Use plugins to associate subject and phenotype data with each variant.
   - Aggregate data into a Cohort Allele Frequency (CAF) object.
3. **Output**:
   - An annotated VCF file including VRS identifiers, subjects, and phenotypes.
   - A JSON CAF object summarizing cohort-level allele frequency data.

---

## Testing

### Unit Tests
- **VCF Parsing**:
  - Validate the ability to read and parse standard VCF files correctly.
  - Ensure compliance with VCF format specifications.
- **VRS Identifier Validation and Generation**:
  - Confirm existing VRS identifiers are recognized.
  - Test the correct generation of new VRS identifiers for variants.
- **Plugin Compliance**:
  - Verify that plugins conform to the interface and return expected data.
- **CAF Generation**:
  - Ensure the cohort allele frequency data is accurately aggregated and formatted.

### Integration Tests
- **End-to-End Workflow**:
  - Test the full pipeline with sample VCF files, including missing VRS identifiers.
  - Confirm that output files (annotated VCF and CAF object) meet specifications.
- **Performance**:
  - Evaluate processing times for large VCF files to ensure scalability.
- **Edge Cases**:
  - Test edge cases such as empty VCF files, malformed lines, or unusual data formats.

### Example Test Case:
1. Input a sample VCF file missing several VRS identifiers.
2. Use plugins to determine subject and phenotype data.
3. Generate a CAF object and validate:
   - Proper VRS identifier assignment.
   - Accurate cohort allele frequency calculation.
   - Correct file outputs (annotated VCF and JSON CAF).

---

**Comment by quinnwai** [comment]
Thanks for the writeup! This is helpful to conceptualize future work. Just to make sure I'm following before diving into discussion:

**Summary**
With a proof of concept CAF generation working for GREGoR, we now want to generalize the usage of the toolkit. The plugin architecture allows a user to process their data according to their own data model/specifics. This handles two cases where...
1. for genotypical data: subjects in VCFs might be for example be stored differently across cohorts
2. for phenotypical data: a project's phenotypical data model might be organized  different across cohorts

Providing multiple access points for the tool (CLI / Python package) will enable multiple researcher use cases 

**Discussion**
- Where would filtering be specified, within a user-implemented plugin? For example, if I want to subset on subjects, that too would be specified in `get_subject`?
- I think the plugins should take in a `pysam.VariantRecord` ([pysam docs](https://pysam.readthedocs.io/en/latest/usage.html#working-with-vcf-bcf-formatted-files)) as input. That way we assume that VCFs are parsed reliably according to VCF conventions.
- While we have not formalized it, our latest work was essentially designing a GREGoR plugin, are we in the business of designing a few more plugins as examples for users? (ties into using 1000G or other open source projects as new tests cases)
- How much do we want to user to specify vs how much do we want to pre-build for them?
  - IOW, how much of our current implementation (VRS-VCF indices, phenotype indexing, etc) we want to include vs how much we want the user to write up as part of their implementation of a plugin.

Overall having the plugin design seems like a perfect fit for our use case though!

**Comment by bwalsh** [comment]
@quinnwai Thanks for the feedback and the discussion today.  As discussed, please review and rewrite as you see fit.  Please forward to James for his comments.

**Comment by quinnwai** [comment]
# Specification: VCF to VRS Processing Package with Cohort Allele Frequency Generation

## Motivation

With a proof of concept Cohort Allele Frequency (CAF) generation working for GREGoR, we now want to generalize this tool for use by a variety of consortia. To get there, collaborators highlighted two important considerations, specifically ensuring...

1. CAF object output is standardized (#94)
2. users can specify how to process data inputs (genotype and phenotype) of different formats

For context, a plugin architecture allows users to process their data according to their own data model. This handles the problem where, for a particular data project...
1. a subjects genotype in a VCF must be aggregated uniquely
    a. Example: Genotype fields in the VCFs are listed by subject IDs, where each subject is a column. These subject IDs must then be used to define a cohort based on subject-level traits.
3. a project's phenotypical data model must be extracted and aggregated uniquely
    a. Example: Rare disease data is stored in an Terra data table, so disease type needs to be organized by subject ID. Alleles should only be counted if a subject has a particular disease type.

Taking into account the state of our current tooling, the following is a high-level design and specification for VRS-compliant CAF object generation.

## Core Features

### 1: VRS Identifier Validation and Generation
- **Input**: Standard VCFs (eg [VCFv4.2](https://samtools.github.io/hts-specs/VCFv4.2.pdf))
- **Functionality**:
  - Check for existing VRS identifiers in each VCF line.
  - Generate VRS identifiers if missing, using:
    - **Input data**: Chromosome, position, reference allele, alternate allele.
    - **Output**: Valid VRS identifiers for each allele according to the [VRS schema](https://vrs.ga4gh.org/).
  - Create a VCF-VRS index to readily access any row in a VCF given any variant expression (HGVS, gnomAD, etc) 
- **Output**: Annotated VCFs, VCF-VRS Index
- [x] Create index using [vrsix](https://github.com/gks-gregor/vrsix)
- [ ] Integrate vrsix functionality into the toolkit
- [ ] Integrate [vrs-python](https://github.com/ga4gh/vrs-python) functionality for on-the-fly VRS ID generation

### 2: Cohort Allele Frequency (CAF) Object Generation
- **Input**:
  - Standard VCFs
  - VCF-VRS index
  - Variant of interest
  - Subcohorts of interest
- **Functionality**:
  - Aggregate allele frequency data for a particular cohort
  - Generate a [CohortAlleleFrequencyStudyResult](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/base-profiles/study-result-profiles.html#cohort-allele-frequency-study-result) object containing:
    - Standardized [VRS](https://vrs.ga4gh.org/en/stable/) variant of interest [`focusAllele` field]
    - Allele frequencies for the cohort (ie the samples from all VCFs) [`focusAlleleCount`, `locusAlleleCount`, and `focusAlleleFrequency` field]
    - Cohort information [`cohort` field, a [StudyGroup](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/core-information-model/entities/study-group.html) object]
    - Data provenance (eg dataset URIs, project of interest, data versioning) [`contributions` field, a list of [Contribution](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/core-information-model/entities/activities/contribution.html#contribution) objects]
    - Additional nested CAFs for a user-specified subcohort (eg demographic/phenotypical stratification of allele frequencies) [`subCohortFrequency`]
- **Output**: CAF object according to pre-release v1.0.0 GA4GH Variant Annotation Specification ([VA-Spec](https://github.com/ga4gh/va-spec/discussions/234))
- [x] Example workflow created in the toolkit for GREGoR with [notebook demo](app.terra.bio/#workspaces/WagnerLab/vrs_anvil/analysis/launch/caf-example.ipynb) on Terra
- [ ] Aggregate across multiple VCFs(?)
- [ ] Modify functionality to output VA-spec [CohortAlleleFrequencyStudyResult](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/base-profiles/study-result-profiles.html#cohort-allele-frequency-study-result) object

### 3: Plugins for Genotype and Phenotype Processing


- **Input**:
  - method to transform patient phenotype information into an index
  - method to apply filters to process each subject's genotype
- **Functionality**:
  - users have greater control over how they want to handle their genotype and phenotype data processing through plugins
  - users can specify a `transform_patient_phenotype` method to process their phenotype data into a patient to phenotype object.
    - **Input:** any set of arguments. This must be flexible to allow. What is more specific is the...
    - **Output:** a dictionary with keys as patient IDs and values as [`MappableConcept`](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/core-information-model/data-types.html#mappableconcept)s
    - Example Use Case: phenotype data should be extracted from specific columns in a Terra data table so that each patient has two characteristics: demographic and disease_type.
  - users can implement a `apply_filters` plugin based on how they want to process each patient's genotype.
    - **Input:** patient to phenotype index, a patient_id, and a genotype
    - **Output:** focus allele count and total allele count
    - Example Use Case (phenotype): alleles should only be counted if a patient is of Asian ancestry with asthma.
    - Example Use Case (genotype): the VCFs used require that haploid variants (chrY, chrM, etc) are processed uniquely
  - otherwise, a user can make use of default plugins that default to processing a data according to the [GREGoR data model](https://gregorconsortium.org/data-model) or adapting them to fit need.
- **Output**: CAF object
- [ ] Restructure toolkit to enable plugin architecture

## Example Workflow

1. **Input**: Provide a VCF file with potentially missing VRS identifiers and a set of subcohorts of interest based on demographics (Asian descent, African descent) or phenotypes (breast cancer, prostate cancer)
2. **Prepare VCF**:
  - Parse the VCF file for existing VRS identifiers and generate any missing ones.
  - Generate VRS-VCF index to rapidly access patient genotype for a given variant
3. **Define Plugins**
  - Use an existing default plugin if your data matches an existing data model or...
  - Implement your own plugins as necessary to:
    - format the phenotype data into a patient-to-phenotype index
    - aggregate according to particular filter 
4. **Generate CAF Object**
  - Use a default or self-implemented plugins to associate subject and phenotype data with each variant.
  - Aggregate data into a Cohort Allele Frequency (CAF) object.
5. **Output**:
  - An annotated VCF file including VRS identifiers, subjects, and phenotypes.
  - A VRS-VCF index to rapidly access patient genotype for a given variant
  - A JSON CAF object summarizing cohort-level allele frequency data.

## Programmatic Interface (Python API)

### Data Types
**[`MappableConcept`](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/core-information-model/data-types.html#mappableconcept)**: Used to store a standardized characteristic about a given cohort (eg demographic information like Asian American or rare phenotype like prostate cancer).

### Functions

**`vcf_to_vrs(input_path: str) -> None`**  
- **Input:**
  - `input_path` (str):
  - `output_path` (str)
- **Functionality:**
  - Annotates all VCFs in the specified path with VRS identifiers
  - Creates a VCF index to enable quick access of said VCFs using any variant ID.
- **Output:** 
  - `annotated_vcf_path` (str)
  - `vrs_vcf_index_path` (str)

**`generate_caf`**
- **Input:**
  - `vcf_paths` (list[str]): list of VCF paths
  - `vrs_vcf_index` (str): Sqlite `.db` file containing the 
  - `variant_id` (str): any variant ID, which will be translated into a VRS ID
  - `phenotypes` (list[MappableConcept]): phenotypes represented as a VA-spec [`MappableConcept`](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/core-information-model/data-types.html#mappableconcept). This will be used to identify cohort allele frequencies for subcohorts in the `subcohortFrequency` field of the outputted [CAF Study Result](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/base-profiles/study-result-profiles.html#cohort-allele-frequency-study-result), as each subcohort (a nested CAF study result) will contain a [StudyGroup](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/core-information-model/entities/study-group.html), which is defined by its `characteristics` field.
  - `plugins` (list[`Plugin`]): list of Plugin type
- **Functionality:**
  - Generates an aggregated allele frequency object for a specified cohort of individuals
  - Allows you to specify additional subcohorts of interest by detailing the characteristics of each subcohort
- **Output:** CAF Study Result object

---

- **`register_plugin(plugin: Plugin) -> None`**  
  Registers a plugin to extend the system’s capabilities.

  **Parameters**:
  - `plugin`: A plugin object implementing the required interface for subject and phenotype determination.


## Testing

### Unit Tests
- **VRS Identifier Validation and Generation**:
  - Confirm existing VRS identifiers can be read from VCF.
  - Test the correct generation of new VRS identifiers for variants.
  - Test VRS-VCF index can be used to index an open access source VCF.
- **Plugin Compliance**:
  - Verify that example plugins conform to the interface (inputs) and return expected data (outputs).
- **CAF Generation**:
  - Test that phenotypes can be readily extracted from open access source data.
  - Ensure the cohort allele frequency data is accurately aggregated and formatted.

### Integration Tests
- **End-to-End Workflow**:
  - Test the full pipeline with sample VCF files and phenotype, including missing VRS identifiers.
  - Test full pipeline on Terra
  - Confirm that output files (annotated VCF and CAF object) meet specifications.
- **Performance**:
  - Evaluate processing times for large VCF files to ensure scalability.

## Discussion Questions

There are two categories of questions we need to answer, some focused on implementation and some on user acceptance...

1. [Implementation] Are these plugins at the right level of granularity? Ideally we want to make a user implement as little as possible in a plugin, yet still have the flexibility to do everything they need and be standardized enough to be consumed downstream.
2. [User Acceptance] One bottleneck possibly preventing users from populating non-active data submissions is that creating an VCF-VRS index takes time and possibly money to process. Even if we have the plugin infrastructure that these users are willing to use, they will still have to devote resource upfront to create new data even if it's just "one more file" (ie the index file). Is there any alternative to this?
3. [User Acceptance] For users to actually make use of a plugin architecture, they will need a way to rapidly test that their plugin is working as expected. Assuming they have the relevant data (index / annotated VCFs / phenotype information), how do they do so?

## List of Action Items

- [ ] Enable CAF generation for multiple VCF files
- [ ] Standardize output to a CAF Study Result object, updating any relevant tests
- [ ] Standardize input of phenotypes as a [`MappableConcept`](https://va-ga4gh.readthedocs.io/en/1.0.0-ballot.2024-11/core-information-model/data-types.html#mappableconcept)
- [ ] Create plugin architecture with the GREGoR data as the default plugin
- [ ] Identify other consortia / data projects to add as an open source example plugin
- [ ] Implement an open source example plugin

**Comment by jsstevenson** [comment]
> One bottleneck possibly preventing users from populating non-active data submissions is that creating an VCF-VRS index takes time and possibly money to process. Even if we have the plugin infrastructure that these users are willing to use, they will still have to devote resource upfront to create new data even if it's just "one more file" (ie the index file). Is there any alternative to this?

Yeah, this is an interesting question and potentially a point of divergence from GREGoR. If cohort frequency generation is, like, a one-off thing, or if you just want to generate a bunch of static CAF objects and be done with it, there's zero reason to generate an index. 

* Maybe index construction can be performed in the same pass as VCF annotation. Depending on the implementation, this might not save a *ton* of compute but it could save some.
* Maybe index construction should be an optional/skippable step 
* If we're talking about a single workflow rather than "file prep" and "CAF generation" workflows, presumably there should be a way to submitted inputs as already-prepared so they don't need to be preprocessed again

**Comment by jsstevenson** [comment]
> 3: Plugins for Genotype and Phenotype Processing

Okay, a few random thoughts, apologies if these are nonsensical

1) it seems easier to map patients to phenotypes (as in HPO/uberon/omim terms) than to map them to demographics. The latter aren't collected into a single controlled vocabulary/ontology as far as I know, which would be important for representing them as MappableConcepts. That's not the end of the world -- users could always supply their own mappings in a plugin -- but it does pose issues from an interoperability/aggregation perspective. Maybe we'd want to take the time to enumerate a list of important terms and make some informed decisions about how to map them, so that we at least establish good practices for the important stuff.

* it's also possible that we could lean on the GREGoR data model as a kind of standard for some of these traits (e.g. sex, but not sure about geographic ancestry).
 
2) I think I don't quite understand how the plugin ecosystem would work. Tell me if if this sounds incorrect: an AnVIL user who wants to extract cohort data from their own bespoke dataset needs to take a gks-anvil supplied WDL workflow, and alter it/employ our plugins (or implement their own) for two tasks, data transformation (i.e. restructuring their data such that the gks-anvil toolkit can map between alleles, patients, and phenotypes/traits) and cohort selection (i.e. providing some traits of interest and only identifying allele frequency for patients who have those traits). Assuming that's more or less correct --

* These seem categorically fairly different. For the "data prep" step, it might be more natural to do shell scripting (awk or bcftools magic) than Python, but the latter kind of sounds like a dataframe or data lookup operation. Not sure if that matters.
* Trait-based filtering seems difficult to predefine, particularly given that it relies on some knowledge of how those traits are identified (see above). Maybe I'm overthinking this though.
